test/performance/odp_cpu_bench.c

```diff
@@ -607,27 +607,28 @@ int main(int argc, char *argv[])
 	printf("first CPU:          %i\n", odp_cpumask_first(&cpumask));
 	printf("cpu mask:           %s\n", cpumaskstr);
 
-	/* Create application queues */
-	if (odp_queue_capability(&queue_capa)) {
-		LOG_ERR("Error: odp_queue_capability() failed\n");
-		exit(EXIT_FAILURE);
+	if (odp_schedule_capability(&schedule_capa)) {
+		printf("Error: Schedule capa failed.\n");
+		return -1;
 	}
 
+	odp_schedule_config(NULL);
```

Comment:
Here's an example of dilemma between capa and config. Capa tells max limits (e.g. queue size 8k), config(NULL) tells that defaults are ok (e.g. queue size 4k), should next capa call after config show 8k or 4k?

4k is now the configured maximum, queue creates over 4k would fail. Although, it would not be very intuitive if capa call before and after config returns different values - although config call is just there for that, change the limits to what's enough for the application.

On the other hand application changes are minimized if config affect capa and applications do config(NULL) *before* capa.

Reply:
Maybe the most logical / easy to use option is to have two capa calls:

// max values for config
odp_schedule_config_capa()

// application configured maximums or defaults
odp_schedule_config()

// run time maximums for e.g. queue_create(). cannot be called before config()
odp_schedule_capa()

Reply:
Then old applications would just call config(NULL) before any schedule calls (including schedule_capa), and would use new schedule_capa() just like old queue_capa.

Reply:
Well, in all other cases capa returns implementation capabilities, no matter what was configured, doesn't it? In this case also capa returns implementation capabilities. It is a task of an application to know what it has configured and which limitations apply because of this configuration.

Reply:
If we're treating config is an access function then by ODP convention `odp_schedule_config()` would be the getter and `odp_schedule_config_set()` would be the setter. That way there is no confusion.

Reply:
typedef struct odp_schedule_config_t {
  uint32_t num_queues;
  uint32_t queue_size;
  uint32_t num_flows;
} odp_schedule_config_t;

typedef struct odp_schedule_capability_t {
  uint32_t max_ordered_locks;
  uint32_t max_groups;
  uint32_t max_prios;
  uint32_t max_queues;
  uint32_t max_queue_size;
  uint32_t max_flows;
  odp_support_t lockfree_queues;
  odp_support_t waitfree_queues;
} odp_schedule_capability_t


// maximums for the config call
int odp_schedule_config_capa(odp_schedule_capability_t *config_capa);

// set config before any other calls
int odp_schedule_config(const odp_schedule_config_t *config);

// run time capability (affected by config)
int odp_schedule_capability(odp_schedule_capability_t *capa);

How this looks like?
1) xxx_config_capa() would output limits for the xxx_config(). Only xxx_config_capa() and xxx_config_init() can be called before xxx_config().
2) xxx_config() must be called always (may accept NULL == all defaults) and before other calls.
3)  xxx_capability() would output limits for other calls (== maximums of the current config)

Reply:
I do not like this approach. I still think that if application uses `odp_schedule_config(NULL)`, it doesn't really care about particular sizes/allowances. If it does, it will fill the config structure as it sees fit. Other approach might be to enforce that max == default.

Reply:
Having defaults be maximum values might not be reasonable either. An implementation-chosen "optimal" value would typically be the default, with capabilities returning the limits that the implementation can support. The application can then either accept the defaults or configure chosen values within the implementation-defined limits.

So `odp_schedule_capability()` might return min, max, and default values for relevant configurable parameters to give the application a complete picture of its range of options. `odp_schedule_config(NULL)` accepts the defaults while `odp_schedule_config_init()` sets parameters to their default values that the application can then override with a subsequent `odp_schedule_config()` call.

The only question remaining is whether the application can query the current scheduler configuration, and if so, how? Since we already have an `odp_schedule_group_info()` API we might want to use that as a model and simply have add `odp_schedule_config_info()` for that purpose.

Reply:
// Max values. Those values not in config_t stay as limits also after config.
// Those values in config_t may change during config_t, use e.g. info_t to check those. 
typedef struct odp_schedule_capability_t {
uint32_t max_ordered_locks;
uint32_t max_groups;
uint32_t max_prios;
uint32_t max_queues;
uint32_t max_queue_size;
uint32_t max_flows;
odp_support_t lockfree_queues;
odp_support_t waitfree_queues;
} odp_schedule_capability_t

// input for _config()
typedef struct odp_schedule_config_t {
uint32_t num_queues;
uint32_t queue_size;
uint32_t num_flows;
} odp_schedule_config_t;

typedef struct odp_schedule_info_t {
uint32_t num_ordered_locks;
uint32_t num_groups;
uint32_t num_prios;
uint32_t num_queues;
uint32_t max_queue_size;
uint32_t num_flows;
bool_t lockfree_queues;
bool_t waitfree_queues;
} odp_schedule_info_t

// Max values for config. Some values are valid over config,
// some not (those included into config_t)
int odp_schedule_capability(odp_schedule_capability_t *capa)

// Set config and start scheduler before any other calls
int odp_schedule_config(const odp_schedule_config_t *config);

// check values of the current config. Some values are equal to other 
// get functions e.g. odp_schedule_num_prio() == info.num_prios
int odp_schedule_info(const odp_schedule_info_t *info);

This is also fine, but it means less trivial changes to current users of queue_capa: 

before:
queue_capa(&capa);
num = capa.sched.max_num

after:
sched_config(NULL);
sched_info(&info);
num = info.num_queues;